### ISSUE TITLE
[RFC]: Unify functional P2P gating — one veto-only verdict for all P2P consumers (NCCL, custom allreduce, symm-mem)

### ISSUE THREAD

**Issue by ldkhang1201** [vllm-project/vllm#51513]
**TL;DR.** On platforms where the driver advertises GPU P2P but the fabric silently drops peer writes (ACS + translated IOMMU, misconfigured BIOS, some VMs), vLLM either hangs forever at engine startup (TP=2). vLLM already owns the right tool — the write-and-verify check `gpu_p2p_access_check` — but today only one of its three (soon four) P2P consumers consults it, it is skipped by default, and it runs *after* the point where NCCL hangs. This RFC proposes to settle the **contract** for functional P2P checking across consumers: (1) make that check the single canonical verdict with **veto-only** semantics, and (2) wire **all** P2P consumers to it, including NCCL via env.

---

### The failure class

`cudaDeviceCanAccessPeer` (and the cuMem capability queries NCCL uses) is a capability lookup — GPU SKU allowlist, chipset allowlist, topology shape — not a data-path test. Whether a peer write actually *arrives* additionally depends on runtime platform state the driver does not model:

- **PCIe ACS** on root ports redirects peer transactions up to the IOMMU;
- an **IOMMU in translated mode** (the default DMA domain on current kernels) has no mapping for peer BAR addresses, so the redirected write faults;
- a PCIe peer write is a **posted** transaction — no acknowledgment exists in the protocol. The dropped write is logged only in the *host* kernel log (`AMD-Vi: IO_PAGE_FAULT` / `DMAR: ... fault`); the GPU and the CUDA stack see nothing. The receiving kernel spin-polls a flag that never arrives.

The result is either an infinite hang (dropped flag writes) or **silent output corruption** (dropped data writes with surviving synchronization).

This is a long-standing, recurring failure class:
- #17676 — "vLLM hangs forever on waiting engine process to start" (open, 42 comments): the umbrella symptom.
- #2728 — the original issue that motivated `can_actually_p2p` ("the driver might be broken, and `can_device_access_peer` returns True even if P2P access is not actually possible").

### Case study — full causal chain, reproduced end to end

Hardware: bare-metal 2-socket AMD EPYC server, 8× PCIe GPUs (no NVLink), one GPU per host bridge/NBIO, Ubuntu kernel 6.17, driver 580.x / CUDA 13.0 / NCCL 2.28.9, workloads in Docker. BIOS ships ACS enabled (each GPU sits alone in its own IOMMU group), kernel default `iommu: Default domain type: Translated`.

1. **The driver blesses every pair.** `nvidia-smi topo -p2p w` reports `OK` for all 64 pairs; `cudaDeviceGetP2PAttribute(AccessSupported, 0, 1) == 1`.

2. **`vllm serve -tp 2` hangs forever** at `Waiting for 1 local, 0 remote core engine proc(s) to start.` with both GPUs pinned at 100%. The stuck frame is the pynccl warmup all-reduce (`vllm/distributed/device_communicators/pynccl.py:142-145`):

   ```python
   stream = current_stream()
   # A small all_reduce for warmup.
   data = torch.zeros(1, device=device)
   self.all_reduce(data)
   stream.synchronize()   # <- never returns; no timeout exists
   ```

3. **The host IOMMU testifies.** Every start attempt logs a burst of faults, stamped with the GPU's BDF, at the moment the warmup spins:

   ```
   nvidia 0000:21:00.0: AMD-Vi: Event logged [IO_PAGE_FAULT domain=0x002e address=0x11b400826200 flags=0x0030]
   nvidia 0000:21:00.0: AMD-Vi: Event logged [IO_PAGE_FAULT domain=0x002e address=0x11b800000000 flags=0x0030]
   ```

   (flags bit 5 = write; the address is the peer GPU's BAR window. The write was redirected by ACS to the IOMMU, which had no translation and dropped it.)

4. **Model-independent minimal repro** (no vLLM involved):

   ```python
   # t.py — hangs on the affected platform at nproc-per-node=2
   import torch, torch.distributed as dist
   dist.init_process_group("nccl")
   r = dist.get_rank(); torch.cuda.set_device(r)
   x = torch.ones(1, device="cuda")
   dist.all_reduce(x); torch.cuda.synchronize()
   print(f"rank {r}: OK sum={x.item()}")
   ```

5. **Why exactly TP=2 (and why most users' "TP=4 works" misleads triage).** NCCL's `ncclTopoCheckP2p` (`src/graph/paths.cc`) defaults to `p2pLevel = PATH_PXB` — *"don't use P2P across CPU Host Bridges"* — which on a switchless topology forbids P2P for every pair. But it carries an AMD-specific carve-out:

   ```c
   if ((arch == NCCL_TOPO_CPU_ARCH_X86 && vendor == NCCL_TOPO_CPU_VENDOR_AMD)
       && system->nodes[DEV].count <= 2)
     p2pLevel = PATH_SYS;
   ```

   Verified empirically on the same pair of GPUs:

   ```
   nranks=2: isAllDirectP2p 1 ... 0[0] -> 1[1] via P2P/CUMEM   -> hangs (timeout kill)
   nranks=4: isAllDirectP2p 0 ... 0[0] -> 1[1] via SHM/direct  -> OK sum=4.0
   nranks=8: isAllDirectP2p 0 ... all hops via SHM/direct      -> OK sum=8.0
   ```

   So NCCL's *distrust* of host-bridge P2P accidentally shields TP≥3, while its AMD ≤2-GPU *trust* carve-out selects exactly the 2-GPU population for the hang. Note this shield covers only NCCL: vLLM's own P2P kernels (below) stay armed at any TP, and host-side `IO_PAGE_FAULT`s were observed during a TP=4 vLLM startup on the same box.

6. **Platform fix exists** (`iommu=pt` or BIOS ACS off — verified: the hang and the faults disappear, P2P works at full speed). But vLLM cannot assume every user can reboot into a fixed platform; its job is to not hang or silently corrupt for the next user who lands on one of these machines.

The complete diagnosis above required host `dmesg` access, `NCCL_DEBUG=INFO`, `py-spy`, and roughly a day — for a failure vLLM already contains the machinery to detect in seconds.

### vLLM today: N consumers, N trust decisions, N kill switches

| P2P consumer | Trust basis today | Functional gate | Kill switch |
|---|---|---|---|
| NCCL / pynccl | NCCL's own topology policy (driver capability + path type) | none | `NCCL_P2P_DISABLE=1` (user must know to set it) |
| `CustomAllreduce` | driver report (`can_device_access_peer`) since #21904 | `gpu_p2p_access_check`, **opt-in** via `VLLM_SKIP_P2P_CHECK=0` | `--disable-custom-all-reduce` |
| `SymmMemCommunicator` | driver report | none | `VLLM_ALLREDUCE_USE_SYMM_MEM=0` |
| MiniMax fused AR (#47544, in flight) | adding its own ad-hoc functional check | own | own |

History: the write-and-verify check ran **by default** until #21904 (`e91d3c9cda`, July 2025) flipped `VLLM_SKIP_P2P_CHECK` to default-skip for startup-cost reasons ("We assume drivers can report p2p status correctly"). Even at its most vigilant setting, today the check protects only custom allreduce — and it executes *after* `PyNcclCommunicator` is constructed (`cuda_communicator.py`), i.e. after the point where the NCCL warmup has already hung. A user who discovers `VLLM_SKIP_P2P_CHECK=0` gains nothing against the hang, and a user who sets `NCCL_P2P_DISABLE=1` un-hangs NCCL only to hand the same broken fabric to custom allreduce and symm-mem (observed: startup proceeds past NCCL, then spins at 100% in the first profiling forward).

Each consumer independently rediscovering this failure mode and growing its own ad-hoc check (#47544 being the latest) is the pattern this RFC wants to end.

### 1. The contract: one canonical verdict, veto-only

Promote `gpu_p2p_access_check(src, tgt)` (subprocess write-and-verify, JSON cache — already implemented in `vllm/distributed/device_communicators/all_reduce_utils.py`) to the single source of truth for "is P2P *functional* between these GPUs".

**Semantics are veto-only:**
- `False` ⇒ P2P is disabled for **every** consumer. A fabric that drops legacy cudaIpc peer writes (ACS/IOMMU/root-complex level) drops cuMem writes too — the veto transfers.
- `True` ⇒ forces nothing. The check exercises legacy cudaIpc mappings, while NCCL 2.28 and torch symmetric memory use cuMem — a pass does not certify those paths, so consumers keep their existing enablement logic.

### 2. Wire all consumers to the verdict

When the check runs (see §3 for when) and returns `False` for any pair in the comm:

- **`CustomAllreduce`** — already wired; unchanged.
- **`SymmMemCommunicator`** — add the same gate before enabling.
- **MiniMax fused AR** — #47544 consumes the shared verdict instead of a private check.
- **NCCL** — vLLM cannot patch libnccl's transport choice, but it owns the worker bootstrap: set `NCCL_P2P_DISABLE=1` in the worker's environment *before the first communicator is created*, unless the user already set any `NCCL_P2P_*` variable themselves (never override explicit user config), and log prominently: `P2P verified non-functional between GPUs 0<->1 (driver reports OK); disabling P2P for NCCL, custom allreduce, and symmetric memory. See <docs link> — likely platform cause: PCIe ACS + IOMMU in translated mode; fix with BIOS ACS off or iommu=pt.`

### 3. Cost story — no new default overhead

- Default behavior is unchanged: check skipped, trusting the driver (per #21904). Healthy platforms pay nothing new.
- With `VLLM_SKIP_P2P_CHECK=0`, the verdict now protects everything, not just custom allreduce — and #39249's async precompute hides its startup latency.
- Follow-up (separate PR, opt-in → maybe default later): auto-enable the check only on risk-class topology — multi-GPU, PCIe-only, no NVLink — and especially the population NCCL's AMD ≤2-GPU carve-out uniquely exposes. NVLink boxes never pay anything.

## CC

Maintainers of the distributed/device-communicator area; author of `can_actually_p2p` and #21904 for the history; #47544 / #45004 / #39249 authors.

---

*Disclosure: this RFC was investigated and drafted with AI assistance (Claude Code). Every claim above was reproduced and verified by the author on real hardware: the dmesg fault signatures, the NCCL transport-selection flip at nranks 2 vs 4/8, the `iommu=pt` fix, and the per-consumer kill-switch behavior. Logs and exact commands available on request. Duplicate-work check performed 2026-08-09: no existing RFC covers unified P2P gating; the related open PRs are point fixes enumerated above.*

**Comment by J-Oe-2** [comment]
Hey everyone, I run a self-hosted lab and hit this hang while bringing up DeepSeek-V4-Flash-0731 on four RTX PRO 6000 Blackwell cards.

The details in this comment were put together by Claude, based on our logs and benchmark reports. So apologies in advance for how it sounds, but it is correct and has all the details I would not have written up myself. So here we go:

An independent reproduction of the TP=2 case this RFC describes, on different hardware and a newer stack, plus one detail about why vLLM's custom all-reduce is enabled at exactly two ranks.

**Environment.** Four NVIDIA RTX PRO 6000 Blackwell Max-Q Workstation Edition cards (sm_120, 98 GB each) on PCIe Gen4 x16, no NVLink, `nvidia-smi topo -m` reporting `NODE` for every pair. AMD host with the AMD-Vi IOMMU in its default translated mode (no `iommu=` kernel options). Both GPUs of the TP=2 pair sit under different root ports, and both root ports report `ACSCtl: SrcValid+ ReqRedir+ CmpltRedir+`. Driver 595.84; inside the container CUDA 13.0, torch 2.13.0+cu130, NCCL 2.29.7; vLLM nightly image at build `0.28.1rc1.dev681+ge7edf17ce`. Model: DeepSeek-V4-Flash-0731, TP=2, `--max-model-len 131072`, `--gpu-memory-utilization 0.96`.

**What we measured.** Each row is a separate boot on the same pair of cards.

| configuration | outcome |
|---|---|
| TP=2, no flags | hangs in the pynccl warm-up all-reduce, no error |
| TP=2, `--disable-custom-all-reduce` only | still hangs |
| TP=2, `NCCL_P2P_DISABLE=1` only | gets past NCCL, KV cache sizes (232,409 tokens), then hangs indefinitely in the FlashInfer autotuning phase: both ranks at 100 % of one CPU core inside a driver module load (`load_binary`, reached from Triton's `_init_handles`); GPUs at 100 % utilisation but 87 W. The kernel logs `AMD-Vi: IO_PAGE_FAULT` against both GPUs in the same second the hang begins, and 49 further faults on those two devices follow. The boot log carries no `Custom allreduce is disabled` warning. |
| TP=2, **both** `NCCL_P2P_DISABLE=1` and `--disable-custom-all-reduce` | serves: first completion after 443 s, `finish_reason: stop`, autotuning completes on both ranks, no new IOMMU faults. NCCL logs `Channel 00 : 0[0] -> 1[1] via SHM/direct` and there is no custom all-reduce initialisation. KV pool 227,198 tokens, i.e. the second flag costs about 2.2 % of pool. |
| TP=4 on the same four cards, no flags | serves; the boot log carries `Custom allreduce is disabled because it's not supported on more than two PCIe-only GPUs`, and the IOMMU fault count did not change across the whole run |

A bare `torch.distributed` NCCL all-reduce with no vLLM involved gives the NCCL half on its own: two ranks hang (on two different pairs of cards), four ranks complete, and two ranks with `NCCL_P2P_DISABLE=1` complete. That is the same minimal reproducer as step 4 of this RFC, so we are not attaching ours unless it would help.

This matches the RFC's causal chain, including its observation that `NCCL_P2P_DISABLE=1` alone only moves the hang into vLLM's own peer-to-peer path. The sequence of the two single-flag attempts is worth noting for triage: each one alone still hangs, so each on its own looks like evidence against the peer-to-peer diagnosis.

**One refinement to the consumer table.** The RFC says vLLM's own P2P kernels stay armed at any TP. For `CustomAllreduce` on a PCIe-only node, the code on main disables it above two ranks and leaves it enabled at exactly two (`vllm/distributed/device_communicators/custom_all_reduce.py`, line 236 at `71888f507a`):

```python
if same_node and world_size > 2 and not fully_connected:
    logger.warning(
        "Custom allreduce is disabled because it's not supported on"
        " more than two PCIe-only GPUs. To silence this warning, "
        "specify disable_custom_all_reduce=True explicitly."
    )
    return
```

At two ranks the only remaining gate is `_can_p2p`, and with the default `VLLM_SKIP_P2P_CHECK=1` (`vllm/envs.py`, line 1195) that returns the driver's `can_device_access_peer` answer, which is OK on this host (`nvidia-smi topo -p2p r` also reports OK). The two-rank exemption is clearly deliberate: `should_custom_ar` (line 418) prefers custom all-reduce at `world_size == 2` on PCIe for performance. So on an AMD PCIe host both hazards are live at exactly two ranks: NCCL's two-GPU AMD carve-out in `paths.cc` that this RFC quotes, and the custom all-reduce exemption. At TP=4 both are off by default, which fits our zero faults at TP=4. We cannot speak to the TP=4 faults the RFC saw on its eight-GPU box. We left the symmetric-memory setting at its default, and the two flags were enough here.

**A scoping control that may help the RFC's risk-class heuristic.** On the same host, under the same IOMMU configuration (we did not check ACS on the 3090s' root ports), a two-rank tensor-parallel vLLM configuration on two RTX 3090s serves normally with neither flag set, and has done so in production. So "two ranks, PCIe-only, AMD, translated IOMMU" is necessary but not sufficient on this box: the Blackwell pair's peer transfers are refused while the Ampere pair's are not. We have not investigated what differs between the two pairs. The practical point is that a topology-only rule would false-positive on a pair like the 3090s, which argues for the functional check over a static heuristic.

**Suggestions, in order of our preference.** These are our proposals, not something we have implemented or tested.
1. The RFC's veto-only verdict, wired to NCCL and custom all-reduce, would have prevented both hangs on this host.
2. As an interim step, when `CustomAllreduce` is about to enable on a pair that is not fully connected and the P2P check is being skipped, run `gpu_p2p_access_check` anyway. That cost falls only on PCIe-only pairs, not on NVLink machines.
3. At minimum, log a warning at that point that names both kill switches, because the failure mode here is a hang with no error at all.
4. Changing the guard from `world_size > 2` to `world_size >= 2` would also avoid the hang, but it would drop custom all-reduce for every healthy two-GPU PCIe host, which line 418 shows is intentional. So we would not suggest it on its own.

**Not tried:**
- `VLLM_SKIP_P2P_CHECK=0` (whether the functional check would have vetoed custom all-reduce in the `NCCL_P2P_DISABLE=1`-only boot).
- `iommu=pt` or clearing ACS redirect. Both need a reboot or root PCI writes, so we have not confirmed the platform fix on this host.
- Other NCCL, driver or vLLM versions.
- Changing the symmetric-memory setting.
- Any throughput comparison between the flag workaround and a platform fix.

Container logs, py-spy stacks and the `dmesg` excerpt for the hung boot are available if useful.

**Comment by ldkhang1201** [comment]
Wonder if some matainer / code owner could take a look, this issue affects all tp2 setup @Isotr0py
